freeipa-client/src/main/java/com/sequenceiq/freeipa/client/FreeIpaClient.java

```diff
@@ -159,11 +165,34 @@ public void userSetPassword(String user, String password) throws FreeIpaClientEx
         // FreeIPA expires any password that is set by another user. Work around this by
         // performing a separate API call to set the password expiration into the future
         userMod(user, "userpassword", password);
-        updateUserPasswordExpiration(user);
+        updateUserPasswordMaxExpiration(user);
+    }
+
+    public void updateUserPasswordExpiration(String user, long expiration) throws FreeIpaClientException {
+        String passwordExpirationDate = formatDate(expiration);
+        userMod(user, "setattr", "krbPasswordExpiration=" + passwordExpirationDate);
+    }
+
+    public void updateUserPasswordMaxExpiration(String user) throws FreeIpaClientException {
+        updateUserPasswordExpiration(user, 0);
     }
 
-    public void updateUserPasswordExpiration(String user) throws FreeIpaClientException {
-        userMod(user, "setattr", "krbPasswordExpiration=" + PASSWORD_EXPIRATION_DATETIME);
+    public User userSetPasswordHash(String user, String hashedPassword, String unencryptedKrbPrincipalKey, long expiration) throws FreeIpaClientException {
+        String passwordExpirationDate = formatDate(expiration);
+        Map<String, Object> params =
+                Map.of("setattr", List.of(
+                        "cdpHashedPassword=" + hashedPassword,
+                        "cdpUnencryptedKrbPrincipalKey=" + unencryptedKrbPrincipalKey,
+                        "krbPasswordExpiration=" + passwordExpirationDate));
+        return userMod(user, params);
+    }
+
+    String formatDate(long timestampMillis) {
+        if (timestampMillis <= 0) {
```

Comment:
I think it's safer if you assert that timestampMillis < 0. It's a bug.

Reply:
will do